jetty-servlets/src/main/java/org/eclipse/jetty/servlets/DoSFilter.java

```diff
@@ -305,67 +307,80 @@ protected void doFilter(HttpServletRequest request, HttpServletResponse response
 
         // Look for the rate tracker for this request.
         RateTracker tracker = (RateTracker)request.getAttribute(__TRACKER);
-        if (tracker == null)
+        if (tracker != null)
+        {
+            // Redispatched, RateTracker present in request attributes.
+            throttleRequest(request, response, filterChain, tracker);
+            return;
+        }
+
+        // This is the first time we have seen this request.
+        if (LOG.isDebugEnabled())
+            LOG.debug("Filtering {}", request);
+
+        // Get a rate tracker associated with this request, and record one hit.
+        tracker = getRateTracker(request);
+
+        // Calculate the rate and check if it is over the allowed limit
+        final boolean overRateLimit = tracker.isRateExceeded(System.currentTimeMillis());
+
+        // Pass it through if we are not currently over the rate limit.
+        if (!overRateLimit)
         {
-            // This is the first time we have seen this request.
             if (LOG.isDebugEnabled())
-                LOG.debug("Filtering {}", request);
+                LOG.debug("Allowing {}", request);
+            doFilterChain(filterChain, request, response);
+            return;
+        }
 
-            // Get a rate tracker associated with this request, and record one hit.
-            tracker = getRateTracker(request);
+        // We are over the limit.
 
-            // Calculate the rate and check if it is over the allowed limit
-            final boolean overRateLimit = tracker.isRateExceeded(System.currentTimeMillis());
+        // So either reject it, delay it or throttle it.
+        long delayMs = getDelayMs();
+        Action wantedAction;
```

Comment:
I wouldn't bother with working out a wantedAction and passing that in. 
The default listener can be made able to see `getDelayMs()` and thus work out the action itself.   We should just delegate policy, not delegate a policy override.